datalad/interface/rerun.py

```diff
@@ -0,0 +1,277 @@
+from datalad.interface.results import get_status_dict
+from datalad.interface.run import run_command
+from datalad.interface.common_opts import save_message_opt
+
+from datalad.support.constraints import EnsureNone, EnsureStr
+from datalad.support.gitrepo import GitCommandError
+from datalad.support.param import Parameter
+
+from datalad.distribution.dataset import require_dataset
+from datalad.distribution.dataset import EnsureDataset
+from datalad.distribution.dataset import datasetmethod
+
+lgr = logging.getLogger('datalad.interface.run')
+
+
+@build_doc
+class Rerun(Interface):
+    """Re-execute previous `datalad run` commands.
+
+    This will unlock any dataset content that is on record to have
+    been modified by the command in the specified revision.  It will
+    then re-execute the command in the recorded path (if it was inside
+    the dataset). Afterwards, all modifications will be saved.
+    """
+    _params_ = dict(
+        revision=Parameter(
+            args=("revision",),
+            metavar="<revision or range>",
+            nargs="?",
+            doc="""re-run command(s) in this revision or range.  This can be a
+            commit-ish that resolves to a single commit whose command
+            should be re-run. Otherwise, it is taken as a revision
+            range, and all the commands that would be shown by `git
+            log <range>` are re-executed.""",
+            default="HEAD",
+            constraints=EnsureStr()),
+        dataset=Parameter(
+            args=("-d", "--dataset"),
+            doc="""specify the dataset from which to rerun a recorded command.
+            If no dataset is given, an attempt is made to identify the dataset
+            based on the current working directory. If a dataset is given,
+            the command will be executed in the root directory of this
+            dataset.""",
+            constraints=EnsureDataset() | EnsureNone()),
+        message=save_message_opt,
+        branch=Parameter(
+            metavar="NAME",
+            args=("-b", "--branch",),
+            doc="create and checkout this branch before rerunning the commands.",
+            constraints=EnsureStr() | EnsureNone()),
+        onto=Parameter(
+            metavar="base",
+            args=("--onto",),
+            doc="""start point for rerunning the commands.  If not specified, commands
+            are executed at HEAD.  This option can be used to specify
+            an alternative start point, which will be checked out with
+            the branch name specified with --branch or in a detached
+            state otherwise.  As a special case, an empty value for
+            this option means to use the parent of the first commit
+            (e.g., 'a' for the range 'a..b' or 'a^' for the revision
+            'a') as the starting point.""",
+            constraints=EnsureStr() | EnsureNone()),
+        root=Parameter(
+            args=("--root",),
```

Comment:
I think the documentation of `revision` should reference this to show that the meaning of a revision could be changed. Also I wonder whether it could be somehow twisted to use `--since` which is the term that is used elsewhere.

Reply:
I kind like to have git-like `revision` specification but indeed, we do have `--since`, and we could have conflict with `--root` removed by using
- `--since=` == `--root`  -- common use case to reproduce the entire history
- `--since=revA` == `revA..HEAD` -- reproduce since some point in the past
- `--since=revA --until=revB` == `revA..revB` - I think this is the rarest usecase to rerun a specific subset of changes? although I do see someone possibly using it to e.g. rerun a sequence of commands off some branch .  I am not sure if `until` is the best word here but it pairs nice with `since` and isn't used anywhere yet (like the initial `--to` I thought to suggest)

Reply:
OK, I can move things over to something along the lines of what
@yarikoptic suggests if you both think that'd be preferable/clearer to
users.  I like the revision ranges and `--root` a la `git rebase`, but
that's not surprising given that's what I chose to go with.

Reply:
> I think the documentation of revision should reference this to show that the meaning of a revision could be changed

I've updated it to do so in case we stick with the current design.

Reply:
@mih what do you think about this?  `revision/--rot` vs `--since + --until`?

Reply:
So, @mih has no opinion and I like mine, so with "both" agree (in that we do not conflict) ;-) !  I like it because I could just say `--since=` to reproduce everything reproducible without seeking for needed rev in the history, and staying "in spirit" with other uses of `--since=` in datalad. And whenever `--since + --until` are specified we just convert to the desired revision range. am I missing any use case I didn't cover with my specification?

Reply:
> I like it because I could just say --since= to reproduce everything reproducible without seeking for needed rev in the history

That's already the case with `--root`.

> and staying "in spirit" with other uses of --since= in datalad. And whenever --since + --until are specified we just convert to the desired revision range

Fair enough.

> am I missing any use case I didn't cover with my specification?

I don't think so.  I'll move things to `--since/--until`.

Reply:
> am I missing any use case I didn't cover with my specification?

What syntax are you thinking of for rerunning a command from a single
revision?  Before it was `rerun REV`.  Now it could be
`rerun --since=REV^ --until=REV` but that is quite long for what I'd
think would be a common case.

It seems like we're trying to reinvent syntax for something git
already has.

Reply:
> What syntax are you thinking of for rerunning a command from a single revision?

My guess is that if --since isn't given at all, just use the revision
pointed to by --until.  But it seems odd to call that "until".

Reply:
You are correct here though that for this use-case it becomes 'awkward'... also I feel that (in both cases) necessity to specify `^` explicitly to point that the commit should be included is not "user friendly"

What about:

- **without any argument/option** -- rerun last commit (if has RUN)
- **REV** -- rerun REV (so it is like the default value here is `HEAD` to fit above case)
- **--since= [REV]** -- rerun from (and including) the first RUN commit [until REV, if specified]
- **--since=REV1 [REV]** -- rerun from (and including!) REV1 [until REV, if specified, otherwise all -- again like to assume REV==HEAD by default]

?

Reply:
actually, may be that "including" notion for `--since=REV1` isn't that great idea since forbids to express myself as `--since=publication` where `publication` is the "stable point".  So indeed not worth making it to "include" it.  But whenever `--since=` it is assume that the first RUN is included.

Reply:
> So indeed not worth making it to "include" it.

Yeah, I don't like the idea of including it for the reason you say.
Also, I think conceptually, it is nice to stick to the idea that
--since=reva and --until=revb maps to reva..revb.  And I don't think
there are many common cases where you'd have to use "^", so I'm not
sure what we'd gain.

> without any argument/option -- rerun last commit (if has RUN)

This is already the case with ce40f78.

> REV -- rerun REV (so it is like the default value here is HEAD to fit above case)

With ce40f78, this would be `rerun --until=REV`, so yeah.. odd.

I can add a REV position argument as well, but since I already dislike
this with ce40f78 and I'd dislike it even more if we were to add the
positional argument on top of --since/--until, worth asking: If we
feel the need to add --since/--until **and** a positional argument, is
this really any simpler/easier for users than the revrange/--root?
(last time I'll object to this move, I swear :])